src/main/drivers/camera_control.c

```diff
@@ -0,0 +1,229 @@
+static uint32_t endTimeMillis;
+
+#ifdef CAMERA_CONTROL_SOFTWARE_PWM_AVAILABLE
+void TIM6_DAC_IRQHandler()
+{
+    IOHi(cameraControlPwm.io);
+
+    TIM6->SR = 0;
+}
+
+void TIM7_IRQHandler()
+{
+    IOLo(cameraControlPwm.io);
+
+    TIM7->SR = 0;
+}
+#endif
+
+void cameraControlInit()
+{
+    if (cameraControlConfig()->ioTag == IO_TAG_NONE)
+        return;
+
+    cameraControlPwm.io = IOGetByTag(cameraControlConfig()->ioTag);
+    IOInit(cameraControlPwm.io, OWNER_CAMERA_CONTROL, 0);
+
+    if (CAMERA_CONTROL_MODE_HARDWARE_PWM == cameraControlConfig()->mode) {
+#ifdef CAMERA_CONTROL_HARDWARE_PWM_AVAILABLE
+        const timerHardware_t *timerHardware = timerGetByTag(cameraControlConfig()->ioTag, TIM_USE_ANY);
+
+        if (!timerHardware) {
+            return;
+        }
+
+        #ifdef USE_HAL_DRIVER
+        IOConfigGPIOAF(cameraControlPwm.io, IOCFG_AF_PP, timerHardware->alternateFunction);
+        #else
+        IOConfigGPIO(cameraControlPwm.io, IOCFG_AF_PP);
+        #endif
+
+        pwmOutConfig(&cameraControlPwm, timerHardware, CAMERA_CONTROL_TIMER_MHZ, CAMERA_CONTROL_PWM_RESOLUTION, 0, 0);
+
+        *cameraControlPwm.ccr = cameraControlPwm.period;
+        cameraControlPwm.enabled = true;
+#endif
+    } else if (CAMERA_CONTROL_MODE_SOFTWARE_PWM == cameraControlConfig()->mode) {
+#ifdef CAMERA_CONTROL_SOFTWARE_PWM_AVAILABLE
+        IOConfigGPIO(cameraControlPwm.io, IOCFG_OUT_PP);
+        IOHi(cameraControlPwm.io);
+
+        cameraControlPwm.period = CAMERA_CONTROL_SOFT_PWM_RESOLUTION;
+        cameraControlPwm.enabled = true;
+
+        NVIC_InitTypeDef nvicTIM6 = {
+            TIM6_DAC_IRQn, NVIC_PRIORITY_BASE(NVIC_PRIO_TIMER), NVIC_PRIORITY_SUB(NVIC_PRIO_TIMER), ENABLE
+        };
+        NVIC_Init(&nvicTIM6);
+        NVIC_InitTypeDef nvicTIM7 = {
+            TIM7_IRQn, NVIC_PRIORITY_BASE(NVIC_PRIO_TIMER), NVIC_PRIORITY_SUB(NVIC_PRIO_TIMER), ENABLE
+        };
+        NVIC_Init(&nvicTIM7);
+
+        RCC->APB1ENR |= RCC_APB1Periph_TIM6 | RCC_APB1Periph_TIM7;
+        TIM6->PSC = 0;
+        TIM7->PSC = 0;
+#endif
+    } else if (CAMERA_CONTROL_MODE_DAC == cameraControlConfig()->mode) {
+        // @todo not yet implemented
+    }
+}
+
+void cameraControlProcess(uint32_t currentTimeUs)
+{
+    if (endTimeMillis && currentTimeUs >= 1000 * endTimeMillis) {
+        if (CAMERA_CONTROL_MODE_HARDWARE_PWM == cameraControlConfig()->mode) {
+            *cameraControlPwm.ccr = cameraControlPwm.period;
+        } else if (CAMERA_CONTROL_MODE_SOFTWARE_PWM == cameraControlConfig()->mode) {
+
+        }
+
+        endTimeMillis = 0;
+    }
+}
+
+static const int cameraPullUpResistance = 47000;
+static const int buttonResistanceValues[] = { 45000, 27000, 15000, 6810, 0 };
+
+static float calculateKeyPressVoltage(const cameraControlKey_e key)
+{
+    const int buttonResistance = buttonResistanceValues[key];
+    return 1.0e-2f * cameraControlConfig()->refVoltage * buttonResistance / (cameraPullUpResistance + buttonResistance);
```

Comment:
This may be precomputed (including ADC_VOLTAGE) and only multiplied with user-supplied compensation factor (refVoltage). Also it should be quite easy to convert to integer math (scaling result  by reasonable amount)